### PR TITLE
Implement the Expected Reciprocal Rank metric

### PR DESCRIPTION
**Pull request by timokau**
## Description

From http://olivier.chapelle.cc/pub/err.pdf. Also includes the more
general "cascade metric" and a plain python implementation of the
abandonment cascade.

## Motivation and Context

The ERR is an interesting addition to our metric collection that can be used in experiments.

## How Has This Been Tested?

I ran the new tests in `tests/metrics.py`.

## Does this close/impact existing issues?

n/a

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by timokau** [comment]
Travis build will be fixed by https://github.com/kiudee/cs-ranking/pull/56.

File: csrank/metrics.py [hunk 1]
```diff
@@ -123,3 +123,79 @@ def topk_acc(y_true, y_pred):
...
+    return cascade_metric(
+        gradings, probability_mapping, utility=lambda r: 1 / r
+    )
+
+
+def cascade_metric(gradings, probability_mapping, utility):
+    """Computes a cascade metric with a given utility function.
+
+    For more details see the original paper at
+    http://olivier.chapelle.cc/pub/err.pdf.
```
Comment (kiudee) [inline csrank/metrics.py]: This could be moved to the References block (see [here](https://github.com/kiudee/cs-ranking/blob/6c4c30cec89363b97d5c2c238cd0a77637b7a4ee/csrank/objectranking/cmp_net.py#L69-L71))

File: csrank/metrics.py [hunk 2]
```diff
@@ -123,3 +123,79 @@ def topk_acc(y_true, y_pred):
...
+    """Maps a grading (from 0 to `max_grade`) to its DCG "gain".
+
+    The gain is defined as in the Discounted Cumulative Gain (DCG)
+    metric. This is always a value in [0, 1] that can be interpreted as
+    a probability.
+    """
+    return (2 ** grading - 1) / (2 ** max_grade)
+
+
+def err(gradings, probability_mapping):
```
Comment (kiudee) [inline csrank/metrics.py]: The other metrics follow the interface:
`def metric(y_true, y_pred):`
Where y_true is a (n_instances, n_objects) shaped array of integers.
Example: A row `[0, 3, 2, 1]` denotes a ranking where the ground truth ordering is A > D > C > B.

Is it possible to unify the metrics with respect to this interface?

File: csrank/metrics.py (hunk 1 above)
Comment (timokau) [reply csrank/metrics.py]: I've made that change, but I'm not sure what citation style to use. For now I've used the MLA style (as given by google scholar).

File: csrank/metrics.py (hunk 2 above)
Comment (timokau) [reply csrank/metrics.py]: Well `err` internally really expects a list of the probabilities that each object is "the relevant one" or "the one the user would stop its search at". So in your example, we might work with

```
[0.8, 0, 0.05, 0.15]
```

Which would imply that `A > D > C > B`.


The way it is currently implemented we use "gradings" instead, so the "ground truth" could for your example could be

| Object | Grading |
| ------ | ------- |
| A      | 4       |
| B      | 1       |
| C      | 2       |
| D      | 3       |

Which would then be transformed to probabilities using the Discounted Cumulative gain:

| Object | Grading | Gain   |
| ------ | ------- | ------ |
| A      | 4       | 0.9375 |
| B      | 1       | 0.0625 |
| C      | 2       | 0.1875 |
| D      | 3       | 0.4375 |

And from here we can actually use `err`. Currently its possible to either pass a different mapping from gradings to probabilities. If we want to standardize on `y_pred` and `y_true`, we'd have to infer the gradings from `y_pred` and then assume some mapping (like dcg).

Should I do that?

I've also added some WIP module documentation that documents the `y_true` and `y_pred` convention.

Comment (kiudee) [reply csrank/metrics.py]: What about a closure, which constructs a metric based on a mapping provided by the user.
Similar to:
```python
def make_ndcg_at_k_loss(k=5):
    def ndcg(y_true, y_pred):
...
```

Comment (timokau) [reply csrank/metrics.py]: I'm not sure what you mean here -- what is `ndcg` and what is `k`?

Comment (timokau) [reply csrank/metrics.py]: I have adapted the tensorflow implementation to the metrics api. Instead of a closure-constructor, the functions can be supplied as parameters with suitable defaults. This way the average user can just use the defaults, while more advanced users can use `functools.partial` to generate different cascade metrics or use different mappings.

The plain-python implementation still needs adaptation, I've hold off on that for now until you've reviewed the tensorflow implementation.

**Comment by timokau** [comment]
The travis failure is because I enabled doctest support in pytest, which causes it to evaluate every file. `csrank/labelranking/placketluce_model.py` uses `autograd` ("Efficiently computes derivatives of numpy code"), which is not in `requirements.txt` and is not tested. So we should either stop using autograd or make it a dependency.

File: csrank/metrics.py (hunk 2 above)
Comment (timokau) [reply csrank/metrics.py]: Similarly, the tests have to be adapted (although the doctests already cover some of the code as well).

Comment (kiudee) [reply csrank/metrics.py]: This implementation of ERR looks good to me.

**Comment by kiudee** [comment]
Indeed, the code in
https://github.com/kiudee/cs-ranking/tree/master/csrank/labelranking
is quite old now and is due for a rewrite.
Since label ranking is not the main focus of the library and the rewrite should come after the migration in v2.x, I would just remove the code from the repository and reintroduce it later.

Automatic differentiation is something Tensorflow and PyTorch are very good at and there is no good reason to have autograd as dependency.
